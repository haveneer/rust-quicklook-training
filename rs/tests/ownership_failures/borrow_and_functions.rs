fn borrow_and_functions() {
    let mut numbers = vec![1, 2, 3, 4, 5];

    println!("The sum is {}",    // Passes reference, 
             consume(&numbers)); // keeps ownership
    println!("The sum is {}",                // Mutable reference,
             add_and_consume(&mut numbers)); // keeps ownership

    println!("{:?}", numbers);
}

fn consume(numbers: &Vec<i32>) -> i32 {
    // numbers is READ-ONLY, cannot be mutated
    numbers.push(42);  //~ error: cannot borrow `*numbers` as mutable
    let sum: i32 = numbers.iter().sum();
    sum
}

fn add_and_consume(numbers: &mut Vec<i32>) -> i32 {
    numbers.push(42);
    consume(numbers)
}

fn main() {}