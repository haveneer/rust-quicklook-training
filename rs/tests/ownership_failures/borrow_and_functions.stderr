error[E0596]: cannot borrow `*numbers` as mutable, as it is behind a `&` reference
  --> tests/ownership_failures/borrow_and_functions.rs:14:5
   |
14 |     numbers.push(42);  //~ error: cannot borrow `*numbers` as mutable
   |     ^^^^^^^^^^^^^^^^ `numbers` is a `&` reference, so the data it refers to cannot be borrowed as mutable
   |
help: consider changing this to be a mutable reference
   |
12 | fn consume(numbers: &mut Vec<i32>) -> i32 {
   |                     ~~~~~~~~~~~~~
